Add initial support for the OpenH264 H.264 software codec

Conan-Kudo opened the pull request:
### Description
This allows users to leverage the OpenH264 codec from Cisco to encode H.264 video content. It is significantly reduced in capability from alternatives, but it does the job.

This also provides a framework for adding support for other H.264 software codecs provided through FFmpeg.

### Motivation and Context
A number of distributions use [OpenH264 distributed by Cisco](http://www.openh264.org/) to provide a fully licensed H.264 implementation (in particular, both [Fedora Linux](https://fedoraproject.org/wiki/OpenH264) and [openSUSE Linux](https://en.opensuse.org/OpenH264) do). However, OBS Studio cannot currently use it.

This change allows OBS Studio to use OpenH264 through FFmpeg if it's available.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
* Tested recording a screencast using OpenH264 + AAC with both Simple and Advanced settings modes
* Tested streaming a screencast using OpenH264 + AAC with both Simple and Advanced settings modes

This was tested on Fedora Linux 38 for x86_64 with ffmpeg 6.0.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented:
Does this require #8015?

Conan-Kudo commented:
> Does this require #8015?

Not yet. This is only the wiring for the ffmpeg plugin, which I think means it only shows up in the "advanced view" right now? I literally just banged this out and am taking a bit of a break before continuing to iterate on it. I've got local builds going so I can test the functionality...

Conan-Kudo commented:
@RytoEX That said, having #8015 land first would make it tremendous easier for the UI wiring I need to do...

Conan-Kudo commented:
I've pulled in #8015 into this pull request to simplify things for me and to show the full context of this change.

Conan-Kudo commented:
I've now tested that this works, thanks to @Gawdl3y's help! 🎉

tytan652 commented:
We recently merged a rework of our CMake build-system.

Now, `plugins/obs-ffmpeg/CMakeLists.txt` and `plugins/obs-ffmpeg/cmake/legacy.txt` should be updated.

tytan652 reviewed (changes requested):
This PR adds FFmpeg OpenH264 because Fedora can't "easily" provide x264.

- But it does not mean that all OBS (or FFmpeg) builds will have OpenH264 support (e.g., Windows, macOS, Flatpak, other distro). There is no reason to have enabled OpenH264 and x264 when the later is easily available.

- In simple mode, the user must not be able to choose between x264 and OpenH264. If x264 is available, OpenH264 should not be shown.
  - If a user loads a profile that was using (in simple mode) OpenH264 and x264 is available, it might be better to make the user switch to x264.

- When it comes to defaults, x264 (if available) should be prioritized over OpenH264.

- Some parts of this PR make the assumption that FFmpeg OpenH264 is always available.
The assumption should be: FFmpeg OpenH264 is always available **if** x264 is not.

tytan652 commented on UI/window-basic-main.cpp:
```diff
@@ -1542,7 +1544,15 @@ bool OBSBasic::InitBasicConfigDefaults()
+
+	bool useX264 = EncoderAvailable("obs_x264");
+
+	if (useX264)
+		config_set_default_string(basicConfig, "AdvOut", "Encoder",
+					  "obs_x264");
+	else
+		config_set_default_string(basicConfig, "AdvOut", "Encoder",
+					  "ffmpeg_openh264");
```
```suggestion
	config_set_default_string(basicConfig, "AdvOut", "Encoder", useX264 ? "obs_x264" : "ffmpeg_openh264");
```

Note: Not formatted

tytan652 commented on UI/window-basic-main.cpp:
```diff
@@ -1674,12 +1683,14 @@ void OBSBasic::InitBasicConfigDefaults2()
 
+	bool useX264 = EncoderAvailable("obs_x264");
+	const char *h264_fallback =
+		(useX264 ? SIMPLE_ENCODER_X264 : SIMPLE_ENCODER_OPENH264);
```
```suggestion
		useX264 ? SIMPLE_ENCODER_X264 : SIMPLE_ENCODER_OPENH264;
```

tytan652 replied:
It's nitpicking, let it unresolved until a another person give his opinion.

Conan-Kudo replied:
Huh, TIL that GitHub won't run CI if there are unresolved conversations if "all conversations must be resolved before merging" is set.

tytan652 replied:
No, it prevents merging but CI runs are not unlimited, so old runs need to finish before running newer.

tytan652 commented on UI/window-basic-settings-stream.cpp:
```diff
@@ -1360,7 +1360,9 @@ static QString get_adv_fallback(const QString &enc)
 	if (enc == "com.apple.videotoolbox.videoencoder.ave.hevc")
 		return "com.apple.videotoolbox.videoencoder.ave.avc";
-	return "obs_x264";
+	if (enc == "obs_x264")
+		return "obs_x264";
+	return "ffmpeg_openh264";
```
I think you misunderstood the function purpose, it fallbacks HEVC encoders to H264 (some improvement needs to be done like you can see in the comment earlier in the code).

Last resort fallback must be obs_x264 if available (and ffmpeg_openh264 if no x264).

tytan652 commented on UI/window-basic-settings-stream.cpp:
```diff
@@ -1583,8 +1587,14 @@ void OBSBasicSettings::ResetEncoders(bool streamOnly)
+		ui->simpleOutStrEncoder->addItem(
+			ENCODER_STR("Software.OpenH264.H264"),
+			QString(SIMPLE_ENCODER_OPENH264));
+	if (service_supports_encoder(vcodecs, "obs_x264"))
```
```suggestion
	else
```

tytan652 commented on UI/window-basic-settings.cpp:
```diff
@@ -5218,10 +5218,18 @@ void OBSBasicSettings::FillSimpleRecordingValues()
+		ui->simpleOutRecEncoder->addItem(
+			ENCODER_STR("Software.OpenH264.H264"),
+			QString(SIMPLE_ENCODER_OPENH264));
+	if (EncoderAvailable("obs_x264")) {
```
```suggestion
	else {
```

derrod commented on plugins/obs-ffmpeg/obs-ffmpeg-h264.c:
```diff
@@ -0,0 +1,237 @@
+	const struct video_output_info *voi = video_output_get_info(video);
+
+	switch (voi->format) {
+	case VIDEO_FORMAT_I420:
+	case VIDEO_FORMAT_NV12:
+		break;
+	default:; // Make the compiler do the right thing
+		const char *const text =
+			obs_module_text("H264.UnsupportedVideoFormat");
+		obs_encoder_set_last_error(encoder, text);
+		blog(LOG_ERROR, "[H.264 encoder] %s", text);
+		return NULL;
+	}
```
This should allow all 8-bit formats, OBS will handle the conversion to I420.

derrod reviewed (changes requested):
As far as I could tell this will also still display the x264 presets in simple mode, which it shouldn't.

Conan-Kudo commented:
> As far as I could tell this will also still display the x264 presets in simple mode, which it shouldn't.

@derrod I've disabled the preset field for OpenH264, thanks for catching that.

Lain-B commented:
Hey @Conan-Kudo, remind me to come back to this after we make our v30 release branch. Sorry about the delay

PatTheMav commented:
I agree with @tytan652 that the PRa has its selection logic backwards: It installs OpenH264 as a fallback for x264, but x264 is a core module and the former is a very distro-specific option.

As such the PR needs to add OpenH264 as an optional codec and check for its availability and still fall back to x264 as the default.

Conan-Kudo commented:
> As such the PR needs to add OpenH264 as an optional codec and check for its availability and still fall back to x264 as the default.

I added it this way because OBS crashes the other way because there will not be x264 in the first place in Fedora's build unless you install the plugin from RPM Fusion.

Conan-Kudo commented:
And ffmpeg is a "core module" too. Both are always present in upstream OBS builds, it depends on detection logic and whether ffmpeg is compiled with OpenH264 support.

PatTheMav commented:
x264 is the built-in fallback and the app is structured around that.

The better way to solve this would be for Fedora to patch `obs-x264` to use OpenH264 so that the program is no wiser.

(Also I’m unsure as to whether that should be implemented in `obs-x264` itself with a specific check for Fedora, because I’m not happy about distro-specific workarounds while we have to maintain support for 3 operating systems already.)

Afaik we officially support Ubuntu and Flatpak and any distro-specific changes need to be made by distributors while packaging the app.

Conan-Kudo commented:
`obs-x264` is libx264 specific, an equivalent for OpenH264 would be to write a whole new plugin for libopenh264. I used ffmpeg because it already has an interface for this, we prefer OpenH264 to be accessed through ffmpeg anyway, and it would relatively straightforward to wire in, based on how AV1 was done.

PatTheMav commented:
Again, that will not work because you are creating a fallback for a fallback, you are not even checking if ffmpeg supports the encoder at runtime before making that decision and with that you also make macOS and Windows fall back to OpenH264 if x264 is not found.

It creates additional code complexity and maintenance burden for an edge case on a single distribution and IMO that’s not a good deal.

PatTheMav commented:
We can add generic support for OpenH264 if it is available (as in "checked if available via FFmpeg" because we do not ship FFmpeg with OpenH264 support on Windows and macOS as we do not support that codec at all), and that availability will probably be limited to Linux only and depends on the distribution-specific variant of FFmpeg available. If that check is positive, it can be added to the list of available encoders. 

But distribution-specific fixes need to be implemented/patched by distributors. Adding a global (as in: affecting all platforms) fallback on OpenH264 if the _actual_ fallback `x264` (which is commonly shipped by OBS Studio itself or expected to be available on the platform) is not a solution IMO.

Upstream distributors can then handle the added complexity of adding an OpenH264 fallback _if_ it is available at all and abort the OBS Studio launch if it cannot be found.

Also AV1 is not a "baseline/fallback" encoder, so the comparison with its implementation is not applicable.

derrod commented:
It would probably be sensible to split the encoder addition and the UI changes into separate PRs so we can move things along. The former is probably fine (with some minor code review nits).

Conan-Kudo commented:
I've split out the first half of this pull request as #10135, per @derrod's request.

DemiMarie commented:
> But distribution-specific fixes need to be implemented/patched by distributors. Adding a global (as in: affecting all platforms) fallback on OpenH264 if the _actual_ fallback `x264` (which is commonly shipped by OBS Studio itself or expected to be available on the platform) is not a solution IMO.

On desktop Linux, OpenH264 _is_ the worst-case fallback.  Software patent restrictions mean that it is the only H.264 codec that all distros can redistribute.

tytan652 commented on UI/data/locale/en-US.ini:
```diff
@@ -1034,7 +1034,8 @@ Basic.Settings.Output.Simple.Warn.Encoder="Warning: Recording with a software en
 Basic.Settings.Output.Simple.Warn.Lossless.Msg="Are you sure you want to use lossless quality?"
 Basic.Settings.Output.Simple.Warn.Lossless.Title="Lossless quality warning!"
-Basic.Settings.Output.Simple.Encoder.Software="Software (x264)"
+Basic.Settings.Output.Simple.Encoder.Software.OpenH264.H264="Software (OpenH264)"
+Basic.Settings.Output.Simple.Encoder.Software.X264.H264="Software (x264)"
```
Avoid having to re-translate because the key is changed, plus OpenH264 will only provide H264.
```suggestion
Basic.Settings.Output.Simple.Encoder.Software.OpenH264="Software (OpenH264)"
Basic.Settings.Output.Simple.Encoder.Software.X264="Software (x264)"
```

tytan652 commented on UI/data/locale/en-US.ini:
```diff
@@ -1045,7 +1046,7 @@ Basic.Settings.Output.Simple.Encoder.Hardware.NVENC.AV1="Hardware (NVENC, AV1)"
 Basic.Settings.Output.Simple.Encoder.Hardware.Apple.H264="Hardware (Apple, H.264)"
 Basic.Settings.Output.Simple.Encoder.Hardware.Apple.HEVC="Hardware (Apple, HEVC)"
-Basic.Settings.Output.Simple.Encoder.SoftwareLowCPU="Software (x264 low CPU usage preset, increases file size)"
+Basic.Settings.Output.Simple.Encoder.SoftwareLowCPU.X264.H264="Software (x264 low CPU usage preset, increases file size)"
```
Avoid having to re-translate because the key is changed.
```suggestion
Basic.Settings.Output.Simple.Encoder.SoftwareLowCPU="Software (x264 low CPU usage preset, increases file size)"
```

tytan652 commented on UI/window-basic-auto-config-test.cpp:
```diff
@@ -981,7 +994,8 @@ void AutoConfigTestPage::TestRecordingEncoderThread()
 
 #define ENCODER_TEXT(x) "Basic.Settings.Output.Simple.Encoder." x
-#define ENCODER_SOFTWARE ENCODER_TEXT("Software")
+#define ENCODER_OPENH264 ENCODER_TEXT("Software.OpenH264.H264")
+#define ENCODER_X264 ENCODER_TEXT("Software.X264.H264")
```
```suggestion
#define ENCODER_OPENH264 ENCODER_TEXT("Software.OpenH264")
#define ENCODER_X264 ENCODER_TEXT("Software")
```

tytan652 commented on UI/window-basic-settings-stream.cpp:
```diff
@@ -1663,7 +1667,11 @@ void OBSBasicSettings::ResetEncoders(bool streamOnly)
 #define ENCODER_STR(str) QTStr("Basic.Settings.Output.Simple.Encoder." str)
 
-	ui->simpleOutStrEncoder->addItem(ENCODER_STR("Software"), QString(SIMPLE_ENCODER_X264));
+	if (!service_supports_encoder(vcodecs, "obs_x264"))
+		ui->simpleOutStrEncoder->addItem(ENCODER_STR("Software.OpenH264.H264"),
+						 QString(SIMPLE_ENCODER_OPENH264));
+	else
+		ui->simpleOutStrEncoder->addItem(ENCODER_STR("Software.X264.H264"), QString(SIMPLE_ENCODER_X264));
```
```suggestion
	if (!service_supports_encoder(vcodecs, "obs_x264"))
		ui->simpleOutStrEncoder->addItem(ENCODER_STR("Software.OpenH264"),
						 QString(SIMPLE_ENCODER_OPENH264));
	else
		ui->simpleOutStrEncoder->addItem(ENCODER_STR("Software"), QString(SIMPLE_ENCODER_X264));
```

tytan652 commented on UI/window-basic-settings.cpp:
```diff
@@ -4735,8 +4740,14 @@ void OBSBasicSettings::FillSimpleRecordingValues()
 
-	ui->simpleOutRecEncoder->addItem(ENCODER_STR("Software"), QString(SIMPLE_ENCODER_X264));
-	ui->simpleOutRecEncoder->addItem(ENCODER_STR("SoftwareLowCPU"), QString(SIMPLE_ENCODER_X264_LOWCPU));
+	if (!EncoderAvailable("obs_x264"))
+		ui->simpleOutRecEncoder->addItem(ENCODER_STR("Software.OpenH264.H264"),
+						 QString(SIMPLE_ENCODER_OPENH264));
+	else {
+		ui->simpleOutRecEncoder->addItem(ENCODER_STR("Software.X264.H264"), QString(SIMPLE_ENCODER_X264));
+		ui->simpleOutRecEncoder->addItem(ENCODER_STR("SoftwareLowCPU.X264.H264"),
+						 QString(SIMPLE_ENCODER_X264_LOWCPU));
```
```suggestion
	if (!EncoderAvailable("obs_x264"))
		ui->simpleOutRecEncoder->addItem(ENCODER_STR("Software.OpenH264"),
						 QString(SIMPLE_ENCODER_OPENH264));
	else {
		ui->simpleOutRecEncoder->addItem(ENCODER_STR("Software"), QString(SIMPLE_ENCODER_X264));
		ui->simpleOutRecEncoder->addItem(ENCODER_STR("SoftwareLowCPU"),
						 QString(SIMPLE_ENCODER_X264_LOWCPU));
```

tytan652 reviewed (changes requested):
Translation nitpik that covers both PR.

Conan-Kudo commented:
This pull request has been rebased on top of current master and has the remaining commits to merge (frontend commits).

Conan-Kudo committed:
frontend: Add support for OpenH264 as the worst-case fallback

OpenH264 exists as the codec of last resort, so it is implemented
such that it is only used as the software codec if x264 is not
available.

Lain-B commented:
Just briefly looking at it, it seems pretty trivial now that everything else has been merged. Anyone else want to give further feedback?

Fenrirthviti commented:
For the fallback part of this, I would prefer if we just removed the entire fallback system in general in favor of a better error reporting that the encoder is missing. I don't want to make our already tenuous and fragile fallback that... really doesn't make sense anymore more complex.

Fenrirthviti commented:
Also, can you please explain why we're intrinsically linking OpenH264's existence to x264 in the first place?

I would prefer that if the encoder is available, it just shows up. If it's not, it doesn't. I'm not sure what value there is in NOT displaying OpenH264 if x264 is available instead of just showing them both?

Fenrirthviti commented:
Thanks, found the comment in one of the old reviews. That is not something we'd prefer, it can be removed, yes.

Conan-Kudo commented:
I dropped it and rebased onto current master.

Conan-Kudo commented:
> For the fallback part of this, I would prefer if we just removed the entire fallback system in general in favor of a better error reporting that the encoder is missing. I don't want to make our already tenuous and fragile fallback that... really doesn't make sense anymore, more complex.

I would prefer we do this separately. I agree with you that the fallback logic is crufty and fragile, but purging it is currently beyond my capabilities. That said, @asahilina has taken the first steps to allow eliminating it with #12326.

Fenrirthviti commented:
I agree that it's a separate change, yes, and outside scope of what this PR is attempting to accomplish. We're doing some internal checking of the relevant code paths to determine if this is a change we're willing to accept short term, and if it is not, have a reasonable alternative proposal.
